sonarqube/resource_sonarqube_qualitygate_usergroup_association.go

```diff
@@ -0,0 +1,176 @@
+package sonarqube
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/satori/uuid"
+)
+
+// GetQualityGateUsergroupAssociation for unmarshalling response body from getting quality gate association
+type GetQualityGateUsergroupAssociation struct {
+	Paging Paging                                       `json:"paging"`
+	Groups []GetQualityGateUsergroupAssociationProjects `json:"groups,omitempty"`
+	Users  []GetQualityGateUsergroupAssociationProjects `json:"users,omitempty"`
+}
+
+// GetQualityGateUsergroupAssociationProjects used in GetQualityGateUsergroupAssociation
+type GetQualityGateUsergroupAssociationProjects struct {
+	Login       string `json:"login,omitempty"`
+	Name        string `json:"name"`
+	Description string `json:"description,omitempty"`
+	Selected    bool   `json:"selected"`
+}
+
+// Returns the resource represented by this file.
+func resourceSonarqubeQualityGateUsergroupAssociation() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceSonarqubeQualityGateUsergroupAssociationCreate,
+		Read:   resourceSonarqubeQualityGateUsergroupAssociationRead,
+		Delete: resourceSonarqubeQualityGateUsergroupAssociationDelete,
+		// Importer: &schema.ResourceImporter{
+		// 	State: resourceSonarqubeQualityGateProjectAssociationImport,
+		// },
+
+		// Define the fields of this schema.
+		Schema: map[string]*schema.Schema{
+			"login_name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ExactlyOneOf: []string{"login_name", "group_name"},
+			},
+			"group_name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ExactlyOneOf: []string{"login_name", "group_name"},
+			},
+			"gatename": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceSonarqubeQualityGateUsergroupAssociationCreate(d *schema.ResourceData, m interface{}) error {
+	sonarQubeURL := m.(*ProviderConfiguration).sonarQubeURL
+	rawQuery := url.Values{
+		"gateName": []string{d.Get("gatename").(string)},
+	}
+
+	if target, ok := d.GetOk("login_name"); ok {
+		sonarQubeURL.Path = strings.TrimSuffix(sonarQubeURL.Path, "/") + "/api/qualitygates/add_user"
+		rawQuery.Add("login", target.(string))
+	} else if target, ok := d.GetOk("group_name"); ok {
+		sonarQubeURL.Path = strings.TrimSuffix(sonarQubeURL.Path, "/") + "/api/qualitygates/add_group"
+		rawQuery.Add("groupName", target.(string))
+	}
+	sonarQubeURL.RawQuery = rawQuery.Encode()
+
+	resp, err := httpRequestHelper(
+		m.(*ProviderConfiguration).httpClient,
+		"POST",
+		sonarQubeURL.String(),
+		http.StatusNoContent,
+		"resourceSonarqubeQualityGateUsergroupAssociationCreate",
+	)
+	defer resp.Body.Close()
+
+	if err != nil {
+		return fmt.Errorf("error creating Sonarqube quality gate association: %+v", err)
```

Comment:
Can you make this error a little more verbose. Just so we know this is for adding a user/group to a quality gate and not the association of the quality gate to something else